lib/winapi/timezoneapi.c

```diff
@@ -0,0 +1,194 @@
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_BIAS, &type, &standardBias, sizeof(standardBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_NAME, &type, dayLightName, sizeof(dayLightName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_NAME, &type, standardName, sizeof(standardName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    if (standardDate.Month > 12 || dayLightDate.Month > 12 || 
+        standardDate.DayOfWeek > 6 || dayLightDate.DayOfWeek > 6 ||
+        standardDate.Hour > 23 || dayLightDate.Hour > 23) {
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    // Xbox stores up to 4 characters here.
+    for (UCHAR i = 0; i < 4; i++) {
+        lpTimeZoneInformation->StandardName[i] = standardName[i];
+        lpTimeZoneInformation->DaylightName[i] = dayLightName[i];
+    }
+
+    // There was no Daylight savings date info
+    if (standardDate.Month == 0 || dayLightDate.Month == 0) {
+        return TIME_ZONE_ID_UNKNOWN;
+    }
+
+    lpTimeZoneInformation->StandardDate.wMonth = standardDate.Month;
+    lpTimeZoneInformation->StandardDate.wDayOfWeek = standardDate.DayOfWeek;
+    lpTimeZoneInformation->StandardDate.wDay = standardDate.Day;
+    lpTimeZoneInformation->StandardDate.wHour = standardDate.Hour;
+    lpTimeZoneInformation->StandardBias = standardBias;
+
+    lpTimeZoneInformation->DaylightDate.wMonth = dayLightDate.Month;
+    lpTimeZoneInformation->DaylightDate.wDayOfWeek = dayLightDate.DayOfWeek;
+    lpTimeZoneInformation->DaylightDate.wDay = dayLightDate.Day;
+    lpTimeZoneInformation->DaylightDate.wHour = dayLightDate.Hour;
+    lpTimeZoneInformation->DaylightBias = dayLightBias;
+
+    // Now we needed to determine if the current time is within the day light savings range
+    KeQuerySystemTime(&kTime);
+    RtlTimeToTimeFields(&kTime, &dateNow);
+
+    // Determine what day in the month DST starts and ends
+    UCHAR endDay = GetActualDay(dateNow.Year, &dayLightDate);
+    UCHAR startDay = GetActualDay(dateNow.Year, &standardDate);
+
+    // Fix month wrapping. i.e if DST is Oct to Feb and current month is Jan we should be in DST.
+    // Month range is 10-2, this becomes 10-14
+    // Current month becomes 13. So we compare 10 < 13 < 14 so it all works
+    if (standardDate.Month < dayLightDate.Month)
+    {
+        if (dateNow.Month <= standardDate.Month) {
+            dateNow.Month += 12;
+        }
+        standardDate.Month += 12;
```

Comment:
It's not clear why `.Month += 12` is used. Wouldn't that mean the Month is out of the 1-12 range? What is being calculated here?

Reply:
We ultimately need to determine if Start of DST < now < End of DST. 

In the case of DST going over new  years may want to calculate if January(1) is between Dec(12) and Feb(2); a simple range check would give us 12<1<2 which is comes out false which is wrong.

The +12 just handles this wrap around business, so ultimately we compare 12<13<14 which is correct.
This feeds into the calculation futher down which also has to consider Days, hours and minutes but only the month wrap around is the main problem.

Effectively, this just considers January the 13th Month  and Feb 14th Month since the start of the previous year

Reply:
Yes, thanks. This is much better :)